request-server_test.go

```diff
@@ -568,27 +568,67 @@ func TestRequestSymlink(t *testing.T) {
 	p := clientRequestServerPair(t)
 	defer p.Close()
 
-	_, err := putTestFile(p.cli, "/foo", "hello")
-	require.NoError(t, err)
+	const CONTENT_FOO = "hello"
+	const CONTENT_DIR_FILE_TXT = "file"
+	const CONTENT_SUB_FILE_TXT = "file-in-sub"
 
-	err = p.cli.Symlink("/foo", "/bar")
+	// prepare all files
+	_, err := putTestFile(p.cli, "/foo", CONTENT_FOO)
 	require.NoError(t, err)
-	err = p.cli.Symlink("/bar", "/baz")
+	err = p.cli.Mkdir("/dir")
+	require.NoError(t, err)
+	err = p.cli.Mkdir("/dir/sub")
+	require.NoError(t, err)
+	_, err = putTestFile(p.cli, "/dir/file.txt", CONTENT_DIR_FILE_TXT)
+	require.NoError(t, err)
+	_, err = putTestFile(p.cli, "/dir/sub/file-in-sub.txt", CONTENT_SUB_FILE_TXT)
 	require.NoError(t, err)
 
-	r := p.testHandler()
+	type symlink struct {
+		name   string // this is the filename of the symbolic link
+		target string // this is the file or directory the link points to
 
-	fi, err := r.lfetch("/bar")
-	require.NoError(t, err)
-	assert.True(t, fi.Mode()&os.ModeSymlink == os.ModeSymlink)
+		//for testing
+		expectsNotExist     bool
+		expectedFileContent string
+	}
 
-	fi, err = r.lfetch("/baz")
-	require.NoError(t, err)
-	assert.True(t, fi.Mode()&os.ModeSymlink == os.ModeSymlink)
+	symlinks := []symlink{
+		{name: "/bar", target: "/foo", expectedFileContent: CONTENT_FOO},
+		{name: "/baz", target: "/bar", expectedFileContent: CONTENT_FOO},
+		{name: "/link-to-non-existent-file", target: "non-existent-file", expectsNotExist: true},
+		{name: "/dir/rel-link.txt", target: "file.txt", expectedFileContent: CONTENT_DIR_FILE_TXT},
+		{name: "/dir/abs-link.txt", target: "/dir/file.txt", expectedFileContent: CONTENT_DIR_FILE_TXT},
+		{name: "/dir/rel-subdir-link.txt", target: "sub/file-in-sub.txt", expectedFileContent: CONTENT_SUB_FILE_TXT},
+		{name: "/dir/abs-subdir-link.txt", target: "/dir/sub/file-in-sub.txt", expectedFileContent: CONTENT_SUB_FILE_TXT},
+		{name: "/dir/sub/parentdir-link.txt", target: "../file.txt", expectedFileContent: CONTENT_DIR_FILE_TXT},
+	}
 
-	content, err := getTestFile(p.cli, "/baz")
-	require.NoError(t, err)
-	assert.Equal(t, []byte("hello"), content)
+	for _, s := range symlinks {
+		err = p.cli.Symlink(s.target, s.name)
```

Comment:
I recommend isolating `err` assignments to as narrow of a scope as possible. So, here is `err := p.cli.Symlink(s.target, s.name)`

It’s far easier to reason about the full scope of an assignment if it’s been well-scoped. For example, with `err =` the value of the assignment here leaks out of the loop… even though we _know_ it has to be `nil` at the conclusion of the loop. By defining a new copy of `err` here, we tell the compiler that this value won’t be leaking out of the loop.

Reply:
Yeah, I normally do that, but missed it here